### ISSUE TITLE
Should the snap offer the option to use APCu for local cache?

### ISSUE THREAD

**Issue by pachulo** [nextcloud-snap/nextcloud-snap#2663]
After reading the [Recommendations based on type of deployment](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#recommendations-based-on-type-of-deployment) on the [Memory caching page of the nextcloud manual](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html), I was wondering:
* Should the snap come with the PHP APCu extension so that it can be used for `memcache.local` instead of Redis?

I haven't found any real benchmarks or anything, just that [it should be faster but consume more RAM](https://wpjohnny.com/redis-vs-memcache-vs-apcu-object-cache-comparisons/), but as it is also the official recommendation, I just wanted to understand better why we use Redis instead.

**Comment by scubamuc** [comment]
@pachulo we have a large user base running "limited memory" devices (SBC's) so this should definitely be optional. 

> I haven't found any real benchmarks or anything, just that [it should be faster but consume more RAM](https://wpjohnny.com/redis-vs-memcache-vs-apcu-object-cache-comparisons/), but as it is also the official recommendation, I just wanted to understand better why we use Redis instead.

According to the recommendation, it could increase performance on "enhanced memory" machines. That would be desirable.

**Comment by kyrofa** [comment]
Yeah, it makes sense that we'd be using redis given our start on the Nextcloud box. However, while I agree with @scubamuc that continuing to support such users is important, I just checked the metrics this morning: literally 95% of our users are amd64. If this would enhance performance, it might be worth investigating.

**Comment by pachulo** [comment]
OK, I found more information:
* In [this nextcloud tunning guide from 2019](https://autoize.com/tuning-nextcloud-performance/) they talk about 2GB being the baseline for this:
  > If you are running on a small cloud instance (say, less than 2GB RAM), or a resource-constrained device such as Raspberry Pi, Redis can be used for both local caching and file locking.
* And they also reference [a (very old) benchmark from](https://github.com/owncloud-archive/pi-image/issues/24) @oparoz :
  > If you have the memory overhead to run both PHP APCu and Redis, you would enjoy better performance by [running APCu as the local cache, and Redis as the file locking cache](https://github.com/owncloud/documentation/pull/1992#issuecomment-203554416) based on a user benchmark on GitHub.


P.D.
Should we also look to move to NGINX based on [this comment](https://autoize.com/tuning-nextcloud-performance/#comment-1092) @kyrofa ? 😝

**Comment by kyrofa** [comment]
> Should we also look to move to NGINX based on [this comment](https://autoize.com/tuning-nextcloud-performance/#comment-1092) @kyrofa ? 😝

No, only [Apache is supported](https://docs.nextcloud.com/server/latest/admin_manual/installation/nginx.html):

> Please note that webservers other than Apache 2.x are not officially supported.

**Comment by scubamuc** [comment]
> checked the metrics this morning: literally 95% of our users are amd64

Only 5% running "limited memory" devices (SBC's), interesting... 

> we have a ~~large~~ user base running "limited memory" devices (SBC's)

I would be remiss if I hadn't mentioned these users?

**Comment by pachulo** [comment]
> @pachulo we have a large user base running "limited memory" devices (SBC's) so this should definitely be optional.

I'm one of such users, so I can't agree more! :smile: 

> Could we configure this based on available RAM?🤔

What do you think would make more sense @kyrofa :
1. to make it an autoconfig, depending on the amount of RAM?;
2. or to create a manual knob?

**Comment by alfi4000** [comment]
Um I am a bit confused I am totally unfamiliar with those stuff so is there a way to tune the server performance because I have 256gb ram 24 core but I had closely a crash on my device that was accessing the Nextcloud instance that is on the high performance server it were looking for me that the access device loads the hole stuff not the server where Nextcloud were running on especially when I uploaded 361 images at once my device from where I was uploading were stuck for 20sec! That was mind blowing for me ! :/

**Comment by alfi4000** [comment]
Any advice would be great! :)

**Comment by scubamuc** [comment]
@alfi4000, you have a powerful machine by the sounds of it... 

this PR is a development discussion and is not the right PR for your issue... I guess. @pachulo is considering some default memory cache changes for the Snap.

as far as your performance issues are concerned, its not as simple as tuning here or there... you'd have to study your system logs and performance monitors carefully to pinpoint the bottleneck. That's admin stuff and not necessarily a Nextcloud snap issue. there may be some knobs to turn if we knew more about your system. but we're trying to make Nextcloud snap work for everyone "out of the box". So we're more concerned with stability than performance... especially since we don't know what device is being used.

basically advice for system tuning your machine would be going way beyond what we're trying to do here... also this PR would be wrong for questions about your device. So I'd suggest you make a new PR with your performance issue.

**Comment by alfi4000** [comment]
Ok

**Comment by user8446** [comment]
What about just including APCu, leaving Redis as the default, and users can simply change `'memcache.local'` in their `config.php` to what they desire?

**Comment by Gabgobie** [comment]
Hi,

this has been dormant for a while. I'd be highly in favor of going with @/user8446 s suggestion. I've been in a somewhat memory constrained environment for a while and was finally able to get a RAM upgrade. I'd really like to enable APCu for the boost in speed.

Outside of maybe SLC USBs I can't really think of an environment where 100kB "bloat" are a deal breaker and would assume that in such a scenario a more tailored installation would be used anyways.

---

The following might be outside the scope of this issue.

Unfortunately giving it more RAM didn't help with performance (at all?) - some pageloads are beyond 10sec. Going through the nextcloud-snap wiki, the only memory-related setting I was able to find was the php memory limit. `free -h` is telling me that the VM currently only makes use of 2.3Gi of memory - significantly less than the PHP memory limit. This is probably expected but I'd appreciate having some ways to make use of the memory to make up for the likely bottleneck, the spinning rust my instance is living on. Potentially in a separate issue we could consider adding more settings in terms of database and Redis memory. I also don't know which setting I can change inside the snap without risking the ***sublime*** stability during updates. I'd assume making changes outside the `snap set nextcloud <setting>=<value>` is unsupported/risky.

Best,
Gab

**Comment by htot** [comment]
> > checked the metrics this morning: literally 95% of our users are amd64
> 
> Only 5% running "limited memory" devices (SBC's), interesting...
> 
> > we have a ~large~ user base running "limited memory" devices (SBC's)
> 
> I would be remiss if I hadn't mentioned these users?

All these terms are weakly defined.  I am using a NUC which is a SBC. _this_ version of NUC has 8GB of RAM. But the processor is amd64 or x86_64. Other NUC's have more or less memory. But my setup seems just fine for home use, when I start an instance of Nextcloud Office I'am using 2GB of RAM. And importantly, for an always on device power consumption is low.

And strictly, all computers have limited memory.

Wouldn't it make more sense to define uses cases? And then you can recommend hardware requirements for the use case.

**Comment by user8446** [comment]
Users can still use redis for local cache if they want by simply defining memcache.local in their config.php
